golem/envs/auto_setup.py

```diff
@@ -0,0 +1,195 @@
+from typing import (
+    Any,
+    Callable,
+    Dict,
+    Optional,
+    Tuple,
+    Type
+)
+
+from twisted.internet.defer import (
+    Deferred,
+    DeferredLock,
+    inlineCallbacks
+)
+
+from golem.envs import (
+    CounterId,
+    CounterUsage,
+    EnvConfig,
+    EnvEventListener,
+    EnvEventType,
+    Environment,
+    EnvMetadata,
+    EnvStatus,
+    EnvSupportStatus,
+    Prerequisites,
+    Runtime,
+    RuntimeEventListener,
+    RuntimeEventType,
+    RuntimeInput,
+    RuntimeOutput,
+    RuntimePayload,
+    RuntimeStatus
+)
+
+
+class RuntimeSetupWrapper(Runtime):
+
+    def __init__(
+            self,
+            runtime: Runtime,
+            start_usage: Callable[[], Deferred],
+            end_usage: Callable[[], Deferred]
+    ) -> None:
+        self._runtime = runtime
+        self._start_usage = start_usage
+        self._end_usage = end_usage
+
+    @inlineCallbacks
+    def prepare(self) -> Deferred:
+        yield self._start_usage()
+        yield self._runtime.prepare()
+
+    @inlineCallbacks
+    def clean_up(self) -> Deferred:
+        yield self._runtime.clean_up()
+        yield self._end_usage()
+
+    def start(self) -> Deferred:
+        return self._runtime.start()
+
+    def wait_until_stopped(self) -> Deferred:
+        return self._runtime.wait_until_stopped()
+
+    def stop(self) -> Deferred:
+        return self._runtime.stop()
+
+    def status(self) -> RuntimeStatus:
+        return self._runtime.status()
+
+    def stdin(self, encoding: Optional[str] = None) -> RuntimeInput:
+        return self._runtime.stdin(encoding)
+
+    def stdout(self, encoding: Optional[str] = None) -> RuntimeOutput:
+        return self._runtime.stdout(encoding)
+
+    def stderr(self, encoding: Optional[str] = None) -> RuntimeOutput:
+        return self._runtime.stderr(encoding)
+
+    def get_port_mapping(self, port: int) -> Tuple[str, int]:
+        return self._runtime.get_port_mapping(port)
+
+    def usage_counters(self) -> Dict[CounterId, CounterUsage]:
+        return self._runtime.usage_counters()
+
+    def listen(
+            self,
+            event_type: RuntimeEventType,
+            listener: RuntimeEventListener
+    ) -> None:
+        self._runtime.listen(event_type, listener)
+
+
+def auto_setup(wrapped: Type[Environment]) -> Type[Environment]:
+
+    class EnvSetupWrapper(Environment):
+
+        def __init__(self, *args, **kwargs) -> None:
+            self._env = wrapped(*args, **kwargs)  # type: ignore
+            self._num_users = 0
+            self._lock = DeferredLock()
+
+        @inlineCallbacks
+        def _start_usage(self) -> Deferred:
+            yield self._lock.acquire()
+            try:
+                if self._num_users < 1:
```

Comment:
i think its more clear to check equal to 0, any reason to use < 1?
```suggestion
                if self._num_users == 0:
```

Reply:
No particular reason. Can change.

Reply:
i dont mind, both works, was just a thought to make it more readable